### PR TITLE
Support skip cached Coprocessor request

### PR DESCRIPTION
**Pull request by breeswish**
###  What have you changed?

This PR implements the [Coprocessor cache](https://github.com/tikv/tikv/issues/4230) protocol for TiKV.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Integration test

### CODE REVIEW

File: src/raftstore/store/region_snapshot.rs [hunk 1]
```diff
@@ -38,13 +42,39 @@ impl RegionSnapshot {
...
 
+    #[inline]
     pub fn get_region(&self) -> &Region {
         &self.region
     }
 
+    #[inline]
+    pub fn get_apply_index(&self) -> Result<u64> {
+        let apply_index = self.apply_index.load(Ordering::SeqCst);
+        if apply_index == std::u64::MAX {
```
Comment (hicqu) [inline src/raftstore/store/region_snapshot.rs]: I prefer to use 0 as default value, it's ok because the min value is 5.

Comment (breeswish) [reply src/raftstore/store/region_snapshot.rs]: Thanks for the suggestion!

**Comment by siddontang** [comment]
PTAL @blacktear23

File: src/storage/kv/mod.rs [hunk 2]
```diff
@@ -123,6 +123,24 @@ pub trait Snapshot: Send + Clone {
...
+    }
+
+    /// Checks whether underlying data is unchanged according to a previously retrieved version.
+    /// If data is unchanged, returns `true`.
+    ///
+    /// If the engine does not support data version, `false` is always returned.
+    #[inline]
+    fn is_data_version_matches(&self, _version: u64) -> bool {
+        false
+    }
```
Comment (sticnarf) [inline src/storage/kv/mod.rs]: What's the purpose of this method? According to the description of `get_data_version`, this method can be implemented by `get_data_version`.

File: src/raftstore/store/region_snapshot.rs [hunk 3]
```diff
@@ -38,13 +42,41 @@ impl RegionSnapshot {
         RegionSnapshot {
             snap,
             region: Arc::new(region),
+            // Use 0 to indicate that we the apply index is missing and we need to KvGet it,
```
Comment (sticnarf) [inline src/raftstore/store/region_snapshot.rs]: Use 0 to indicate that <del>we</del> the apply index is missing and we need to KvGet it,

File: src/coprocessor/endpoint.rs [hunk 4]
```diff
@@ -293,7 +304,18 @@ impl<E: Engine> Endpoint<E> {
                     .map(|_| (tracker, snapshot))
             })
             .and_then(move |(tracker, snapshot)| {
-                future::result(handler_builder(snapshot, &tracker.req_ctx))
+                let mut builder = handler_builder;
+
+                if let Some(match_version) = tracker.req_ctx.cache_match_version {
```
Comment (lonng) [inline src/coprocessor/endpoint.rs]:
```rust
if tracker.req_ctx.cache_match_version == snapshot.get_data_version() {
    builder = CachedRequestHandler::builder();
}
```

Comment (sticnarf) [reply src/coprocessor/endpoint.rs]: we shouldn't return cache if they are two `None`s .

**Pull request merged**